### PR TITLE
Add Ddoc headers for front-end modules

### PR DESCRIPTION
**Pull request by dkorpel**
As suggested by Walter: https://github.com/dlang/dmd/pull/9844#issuecomment-495951635

- Adds missing module descriptions
- Consistently puts the description above the boilerplate
- Adds links to the specification where applicable

The descriptions are sometimes vague since I'm not always sure myself what each module does exactly, but it's a start.

### CODE REVIEW

**Comment by Geod24** [comment]
I would just remove the boilerplate. Especially with the way you put it, it's part of the "Specification" section, not the description.

File: src/dmd/arrayop.d [hunk 1]
```diff
@@ -1,4 +1,8 @@
 /**
+ * Implement array opterations, such as `a[] = b[] + c[]`.
```
Comment (Geod24) [inline src/dmd/arrayop.d]: Typo

File: src/dmd/ast_node.d [hunk 2]
```diff
@@ -1,4 +1,6 @@
 /**
+ * Defines an abstract ASTNode class.
```
Comment (Geod24) [inline src/dmd/ast_node.d]:
```suggestion
 * Defines the base class for all nodes which are part of the AST.
```

File: src/dmd/astbase.d [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/astbase.d, _astbase.d)
  * Documentation:  https://dlang.org/phobos/dmd_astbase.html
  * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/astbase.d
  */
 
+module dmd.astbase;
+
+/**
+ */
```
Comment (Geod24) [inline src/dmd/astbase.d]: ?

File: src/dmd/attrib.d [hunk 4]
```diff
@@ -1,4 +1,17 @@
 /**
+ * Defines declarations of various 'attributes'.
+ * The term attribute is used very loosly, since it includes:
```
Comment (Geod24) [inline src/dmd/attrib.d]: Put an empty line (well, with a `*`) between line 2 and 3 to make this part of the description.
Also I wouldn't say it is not used loosely, just that your expectation of what an attribute is might differ.
It has more to do with things applying to a larger scope than a single declaration, I'd say.

File: src/dmd/builtin.d [hunk 5]
```diff
@@ -1,4 +1,7 @@
 /**
+ * Evaluate certain external functions for CTFE.
```
Comment (Geod24) [inline src/dmd/builtin.d]:
```suggestion
 * Implement CTFE for intrinsic (builtin) functions
```
Also the empty line to separate title from description.
You might want to link / read https://en.wikipedia.org/wiki/Intrinsic_function

File: src/dmd/cli.d [hunk 6]
```diff
@@ -1,12 +1,12 @@
 /**
- * Compiler implementation of the
- * $(LINK2 http://www.dlang.org, D programming language).
- *
  * This modules defines the help texts for the CLI options offered by DMD.
```
Comment (Geod24) [inline src/dmd/cli.d]: Newline after this to make the title short.

File: src/dmd/dmangle.d [hunk 7]
```diff
@@ -1,4 +1,9 @@
 /**
+ * Does name mangling for D symbols, which generates a unique name for a linker symbol based
```
Comment (Geod24) [inline src/dmd/dmangle.d]: Too long, and inconsistent with the `cppmangle` side which does not explain what mangling is.
But really, I would expect anyone working on a compiler to know that. We can still add a `See_Also` to wikipedia if you feel like it, but we shouldn't paraphrase it.

**Review by Geod24 — COMMENTED** [review]
Included few comments (didn't finish my review).

In general:
- Get rid of the boilerplate
- Make sure you have an empty line between the title and the description.
- Make sure the descriptions comes directly after the title, not after a section.

File: src/dmd/attrib.d [hunk 8]
```diff
@@ -1,10 +1,12 @@
 /**
- * Defines declarations of various 'attributes'.
- * The term attribute is used very loosly, since it includes:
+ * Defines declarations of various attributes.
+ *
+ * The term 'attribute' refers to things that apply to a larger scope than a single declaration.
```
Comment (Geod24) [inline src/dmd/attrib.d]:
```suggestion
 * The term 'attribute' refers to things that can apply to a larger scope than a single declaration.
```

E.g. depending on how you declare it, you might have `deprecated` as an attribute which is a parent of Dsymbols, or directly as part of the Dsymbol's storage classes.

File: src/dmd/dmangle.d [hunk 9]
```diff
@@ -1,6 +1,5 @@
 /**
- * Does name mangling for D symbols, which generates a unique name for a linker symbol based
- * on the namespace, identifier and type.
+ * Does name mangling for D symbols.
```
Comment (Geod24) [inline src/dmd/dmangle.d]:
```suggestion
 * Does name mangling for `extern(D)` symbols.
```

**Comment by dkorpel** [comment]
> I would just remove the boilerplate. Especially with the way you put it, it's part of the "Specification" section, not the description.

Do you mean this text:

> Compiler implementation of the $(LINK2 http://www.dlang.org, D programming language)

Or the other links as well? Except all for 'Authors', all default sections could be auto-generated, but I do think Walter likes to have them in the source file.

**Comment by Geod24** [comment]
> Do you mean this text:

Yep just this. The rest is fine.

**Review by Geod24 — APPROVED** [review]

**Pull request merged**